apps/extension/client/src/checkTypegenNesting.ts

```diff
@@ -0,0 +1,150 @@
+import * as vscode from "vscode";
+import { WorkspaceConfiguration } from "vscode";
+
+const typeGenPatternKey = "*.ts";
+const typeGenPattern = "${capture}.typegen.ts";
+
+// Checks if the user has our typegen pattern set in their workspace config
+const hasTypeGenPattern = (fileNestingPatterns: object) =>
+  Object.values(fileNestingPatterns).some((pattern: string) =>
+    pattern
+      .split(",")
+      .map((part) => part.trim())
+      .some((part) => part === typeGenPattern)
+  );
+
+const enableTypeGenNesting = (
+  fileNestingConfig: WorkspaceConfiguration,
+  fileNestingPatterns: object
+) => {
+  // If the user chooses to enable file nesting for typegen files we need to make sure that VSCode's file nesting is also enabled
+  fileNestingConfig.update("enabled", true, true);
+
+  // Check if the user has a pattern defined for ts files
+  const existingTsPattern = fileNestingPatterns[typeGenPatternKey];
+
+  // If the user has defined a pattern for ts files we add our typegen pattern to it, otherwise we create a new pattern
+  const updatedPattern = existingTsPattern
+    ? `${existingTsPattern}, ${typeGenPattern}`
+    : typeGenPattern;
+
+  // Update file nesting patterns with all existing patterns and our updated pattern
+  fileNestingConfig.update(
+    "patterns",
+    {
+      ...fileNestingPatterns, // Reuse existing patterns
+      [typeGenPatternKey]: updatedPattern, // Add our updated pattern
+    },
+    true
+  );
+};
+
+const disableTypeGenNesting = (
+  xstateConfig: WorkspaceConfiguration,
+  fileNestingConfig: WorkspaceConfiguration,
+  fileNestingPatterns: object
+) => {
+  xstateConfig.update("nestTypegenFiles", false, true);
```

Comment:
I am somewhat confused by this line - especially when it's called from within `onDidChangeConfiguration`. If a configuration just has been changed - why do we need to update it again here? I see the third argument here that makes it to change the global config, but what's that? 🤔

Reply:
This was mainly done so that we could use the same helper in the three places it's called. But our call to `xstateConfig.update("nestTypegenFiles", false, true)` was actually just useful in one of those places, so I moved it there.

Reply:
Regarding the third argument to `update`, setting this to true will update the global settings (in my case this is my synced user settings). If this is false it will just update the workspace settings. My take is that this config is probably something a user wants to set globally as the default, if not they can still manually set their workspace settings.